### PR TITLE
FastProtocolBridge

### PR DESCRIPTION
**Pull request by swankjesse**
This is a new implementation of ProtocolBridge that uses Zipline's new asDynamicFunction() API. It's dramatically faster - in one measurement we went from spending ~18% of our samples on serialization to ~3% of our samples.

### CODE REVIEW

File: redwood-treehouse-guest/src/jsMain/kotlin/app/cash/redwood/treehouse/guest/FastProtocolBridge.kt [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+  widgetSystemFactory: ProtocolWidgetSystemFactory,
+  private val mismatchHandler: ProtocolMismatchHandler = ProtocolMismatchHandler.Throwing,
+) : ProtocolBridge {
+  private var nextValue = Id.Root.value + 1
+  private val widgets = JsMap<Int, ProtocolWidget>()
+  private var changes = JsArray<Change>()
+  private lateinit var changesSinkService: ChangesSinkService
+  private lateinit var sendChanges: (service: ChangesSinkService, args: List<*>) -> Any?
+
+  public override val widgetSystem: WidgetSystem<Unit> =
```
Comment (swankjesse) [inline redwood-treehouse-guest/src/jsMain/kotlin/app/cash/redwood/treehouse/guest/FastProtocolBridge.kt]: Too much duplication here!

**Review by JakeWharton — APPROVED** [review]

File: redwood-treehouse-guest/src/jsMain/kotlin/app/cash/redwood/treehouse/ProtocolBridgeJs.kt [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+internal actual fun ProtocolBridge(
+  json: Json,
+  hostVersion: RedwoodVersion,
+  widgetSystemFactory: ProtocolWidgetSystemFactory,
+  mismatchHandler: ProtocolMismatchHandler,
+): ProtocolBridge = FastProtocolBridge(json, hostVersion, widgetSystemFactory, mismatchHandler)
+
+@OptIn(RedwoodCodegenApi::class)
+private class FastProtocolBridge(
```
Comment (JakeWharton) [inline redwood-treehouse-guest/src/jsMain/kotlin/app/cash/redwood/treehouse/ProtocolBridgeJs.kt]: I super dislike this name, but I super don't care when `private`!

Comment (swankjesse) [reply redwood-treehouse-guest/src/jsMain/kotlin/app/cash/redwood/treehouse/ProtocolBridgeJs.kt]: Can fix!

File: redwood-treehouse-guest/src/jsTest/kotlin/app/cash/redwood/treehouse/FastProtocolBridgeTest.kt [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.decodeFromDynamic
+import kotlinx.serialization.modules.SerializersModule
+
+/**
+ * Confirm that [FastProtocolBridge] behaves the same as [DefaultProtocolBridge].
+ */
+@OptIn(ExperimentalSerializationApi::class, RedwoodCodegenApi::class)
+class FastProtocolBridgeTest {
```
Comment (swankjesse) [inline redwood-treehouse-guest/src/jsTest/kotlin/app/cash/redwood/treehouse/FastProtocolBridgeTest.kt:47]: FYI @JakeWharton this is new!

File: redwood-protocol-guest/src/commonMain/kotlin/app/cash/redwood/protocol/guest/ProtocolBridge.kt [hunk 4]
```diff
@@ -64,38 +65,59 @@ public interface ProtocolBridge : EventSink {
...
+   * here by special casing that type.
+   *
+   * https://github.com/Kotlin/kotlinx.serialization/issues/2713
+   */
+  @RedwoodCodegenApi
+  public fun appendPropertyChange(
+    id: Id,
+    tag: PropertyTag,
+    value: UInt,
+  )
```
Comment (swankjesse) [inline redwood-protocol-guest/src/commonMain/kotlin/app/cash/redwood/protocol/guest/ProtocolBridge.kt:104]: This is new! And it’s quite unfortunate!

I don’t think we need to implement something general purpose for other kotlinx.serialization types. UInt is something we’re already using in Cash App, so we’re pinned here.

**Pull request merged**